lib/ansible/modules/cloud/amazon/ec2_vpc_endpoint.py

```diff
@@ -0,0 +1,384 @@
+
+    elif module.params.get('policy_file'):
+        try:
+            with open(module.params.get('policy'), 'r') as json_data:
+                policy = json.load(json_data)
+        except Exception as e:
+            module.fail_json(msg=str(e), exception=traceback.format_exc())
+
+    if policy:
+        params['PolicyDocument'] = json.dumps(policy)
+
+    try:
+        changed = True
+        result = camel_dict_to_snake_dict(client.create_vpc_endpoint(**params)['VpcEndpoint'])
+        if token_provided and (request_time > result['creation_timestamp'].replace(tzinfo=None)):
+            changed = False
+        elif module.params.get('wait') and not module.check_mode:
+            status_achieved, result = wait_for_status(client, module, result['vpc_endpoint_id'], 'available')
+            if not status_achieved:
+                module.fail_json(msg='Error waiting for vpc endpoint to become available - please check the AWS console')
+    except botocore.exceptions.ClientError as e:
+        if "DryRunOperation" in e.message:
+            changed = True
+            result = 'Would have created VPC Endpoint if not in check mode'
+        elif "IdempotentParameterMismatch" in e.message:
+            module.fail_json(msg="IdempotentParameterMismatch - updates of endpoints are not allowed by the API")
+        elif "RouteAlreadyExists" in e.message:
+            module.fail_json(msg="RouteAlreadyExists for one of the route tables - update is not allowed by the API")
+        else:
+            module.fail_json(msg=str(e), exception=traceback.format_exc())
+    except Exception as e:
+        module.fail_json(msg=str(e), exception=traceback.format_exc())
+
+    return changed, result
+
+
+def setup_removal(client, module):
+    params = dict()
+    changed = False
+    params['DryRun'] = module.check_mode
+    if not module.params.get('vpc_endpoint_id'):
+        module.fail_json(msg='vpc_endpoint_id is a required paramater')
+    else:
+        if isinstance(module.params.get('vpc_endpoint_id'), basestring):
+            params['VpcEndpointIds'] = [module.params.get('vpc_endpoint_id')]
+        else:
+            params['VpcEndpointIds'] = module.params.get('vpc_endpoint_id')
+    try:
+        result = client.delete_vpc_endpoints(**params)['Unsuccessful']
+        if not module.check_mode and (result != []):
+            module.fail_json(msg=result)
+    except botocore.exceptions.ClientError as e:
+        if "DryRunOperation" in e.message:
+            changed = True
+            result = 'Would have deleted VPC Endpoint if not in check mode'
+        else:
+            module.fail_json(msg=str(e), exception=traceback.format_exc())
+    except Exception as e:
+        module.fail_json(msg=str(e), exception=traceback.format_exc())
+
+    return changed, result
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            vpc_id=dict(),
```

Comment:
required_if is completely undocumented. The module developer guidelines really need an update.